core/src/avm2/verify.rs

```diff
@@ -0,0 +1,1149 @@
+    // while this implementation throws errors #1011 or #1021 in those cases.
+    let mut reader = Reader::new(&body.code);
+    loop {
+        let op = match reader.read_op() {
+            Ok(op) => op,
+
+            Err(AbcReadError::InvalidData(_)) => {
+                return Err(Error::AvmError(verify_error(
+                    activation,
+                    "Error #1011: Method contained illegal opcode.",
+                    1011,
+                )?));
+            }
+            Err(AbcReadError::IoError(_)) => break,
+            Err(unknown) => {
+                tracing::error!(
+                    "Encountered unexpected error while verifying AVM2 method body: {unknown:?}"
+                );
+                break;
+            }
+        };
+        let byte_offset = reader.as_slice().as_ptr() as usize - body.code.as_ptr() as usize;
+
+        new_code.push(op);
+
+        byte_offset_to_idx.insert(byte_offset as i32, new_code.len() as i32);
+        idx_to_byte_offset.push(byte_offset as i32);
+    }
+
+    // Avoid verifying the same blocks again.
+    let mut verified_blocks = Vec::new();
+
+    verify_code_starting_from(
+        activation,
+        method,
+        idx_to_byte_offset.as_slice(),
+        &byte_offset_to_idx,
+        &mut verified_blocks,
+        new_code.as_slice(),
+        0,
+    )?;
+
+    // Record a list of possible places the code could
+    // jump to- this will be used for optimization.
+    let mut potential_jump_targets = HashSet::new();
+
+    // Handle exceptions
+    let mut new_exceptions = Vec::new();
+    for exception in body.exceptions.iter() {
+        // FIXME: This is actually wrong, we should be using the byte offsets, not the opcode offsets.
+        // avmplus allows for from/to (but not targets) that aren't on a opcode, and some obfuscated
+        // SWFs have them. FFDEC handles them correctly, stepping forward byte-by-byte until it
+        // reaches the next opcode. This does the same (stepping byte-by-byte), but it would be better
+        // to directly use the byte offsets when handling exceptions.
+        let mut from_offset = byte_offset_to_idx
+            .get(&(exception.from_offset as i32))
+            .copied();
+
+        let mut offs = 0;
+        while from_offset.is_none() {
+            from_offset = byte_offset_to_idx
+                .get(&((exception.from_offset + offs) as i32))
+                .copied();
+            offs += 1;
+            if offs as usize >= new_code.len() {
+                return Err(make_error_1054(activation));
+            }
+        }
+
+        // Now for the `to_offset`.
+        let mut to_offset = byte_offset_to_idx
+            .get(&(exception.to_offset as i32))
+            .copied();
+
+        let mut offs = 0;
+        while to_offset.is_none() {
+            to_offset = byte_offset_to_idx
+                .get(&((exception.to_offset + offs) as i32))
+                .copied();
+            if offs == 0 {
```

Comment:
`offs` is initialized to 0, so this branch will be taken on the very first iteration through the loop.